### ISSUE TITLE
knife-server with vagrant: can't convert Array into String (TypeError)

### ISSUE THREAD

**Issue by morkot** [fnichol/knife-server#50]
$ knife server bootstrap standalone --node-name chef --host 192.168.33.10 --ssh-user vagrant --ssh-password vagrant --verbose -V
/usr/lib/ruby/gems/1.9.1/gems/net-ssh-2.7.0/lib/net/ssh/transport/session.rb:67:in `initialize': can't convert Array into String (TypeError)
        from /usr/lib/ruby/gems/1.9.1/gems/net-ssh-2.7.0/lib/net/ssh/transport/session.rb:67:in`open'
        from /usr/lib/ruby/gems/1.9.1/gems/net-ssh-2.7.0/lib/net/ssh/transport/session.rb:67:in `block in initialize'
        from /usr/lib/ruby/1.9.1/timeout.rb:55:in`timeout'
        from /usr/lib/ruby/1.9.1/timeout.rb:100:in `timeout'
        from /usr/lib/ruby/gems/1.9.1/gems/net-ssh-2.7.0/lib/net/ssh/transport/session.rb:67:in`initialize'
        from /usr/lib/ruby/gems/1.9.1/gems/net-ssh-2.7.0/lib/net/ssh.rb:200:in `new'
        from /usr/lib/ruby/gems/1.9.1/gems/net-ssh-2.7.0/lib/net/ssh.rb:200:in`start'
        from /usr/lib/ruby/gems/1.9.1/gems/knife-server-1.1.0/lib/knife/server/ssh.rb:46:in `exec!'
        from /usr/lib/ruby/gems/1.9.1/gems/knife-server-1.1.0/lib/chef/knife/server_bootstrap_standalone.rb:83:in`check_ssh_connection'
        from /usr/lib/ruby/gems/1.9.1/gems/knife-server-1.1.0/lib/chef/knife/server_bootstrap_standalone.rb:47:in `run'
        from /usr/lib/ruby/gems/1.9.1/gems/chef-11.8.0/lib/chef/knife.rb:485:in`run_with_pretty_exceptions'
        from /usr/lib/ruby/gems/1.9.1/gems/chef-11.8.0/lib/chef/knife.rb:174:in `run'
        from /usr/lib/ruby/gems/1.9.1/gems/chef-11.8.0/lib/chef/application/knife.rb:133:in`run'
        from /usr/lib/ruby/gems/1.9.1/gems/chef-11.8.0/bin/knife:25:in `<top (required)>'
        from /usr/bin/knife:23:in`load'
        from /usr/bin/knife:23:in `<main>'

$ ssh vagrant@192.168.33.10
vagrant@192.168.33.10's password:
Welcome to Ubuntu 12.04 LTS (GNU/Linux 3.2.0-23-generic x86_64)
- Documentation:  https://help.ubuntu.com/
  Welcome to your Vagrant-built virtual machine.
  Last login: Thu Jan 30 23:24:18 2014 from 10.0.2.2
  vagrant@precise64:~$ logout
  Connection to 192.168.33.10 closed.

**Comment by n1koo** [comment]
Seems to be an issue with starting with Chef 11.8.0. Works fine with 11.6.2 but after migration to anything after that I get the same error.

**Comment by johnbellone** [comment]
I have same issue. Any luck @n1koo ?

**Comment by johnbellone** [comment]
So I figured this out. 

This is what's going into the SSH connection. It seems the defaults are no longer suitable (it is set as an array). We need to explicitly set the `ssh_port`, `ssh_user` and `ssh_password`.

``` ruby
{:user=>"root", :port=>[:ssh_port, nil], :host=>"100.125.1.86", :password=>[:ssh_password, nil], :keys=>["/root/.ssh/root.pem"]}
```

``` ruby
{:user=>"root", :port=>"22", :host=>"100.125.1.86", :password=>nil, :keys=>["/root/.ssh/root.pem"]}
```

**Comment by rberger** [comment]
So how does one actually fix this? I don't have a hash like @johnbellone shows in my config? Is this a bug in the code?
Thanks!

**Comment by arosenhagen** [comment]
#48 fixes this issue and should be merged.

**Comment by n1koo** [comment]
I think this is the most elegant fix: https://github.com/johnbellone/knife-server/commit/d37e9ec4144f03e5983b32648c71a9e20fd224d3

Although it isn't perfect either. The keys array isn't empty with the default parameters, so I made an ugly glue for myself

 'DEFAULT_OPTIONS = { :user => "root", :port => "22", :keys => [] }.freeze' in ssh.rb

EDIT: The actual problem seems to lie in  Chef::Config[:knife].fetch which returns a hash for some reason. I "fixed" it temporarily by replacing it with  Chef::Config[:knife][:key] but then you lose the default values.

**Comment by ScottBouloutian** [comment]
I am experiencing this problem as well. I cannot get knife-server to run. I followed all the instructions and every time it gives me the error:
ERROR: TypeError: can't convert Array into String

Any updates? I would love to use knife-server.

**Comment by fnichol** [comment]
@n1koo good sleuthing, yes it looks like there was a funky change in `Chef::Config[:knife]#fetch` behavior. Sadly I didn't split this fix out by itself, but is in f4705ac1e464ef9fa7b0f21c62b4c6840b394dab (https://github.com/fnichol/knife-server/commit/f4705ac1e464ef9fa7b0f21c62b4c6840b394dab#commitcomment-7650041).

I'm circling back on #48 with a rebase and merge now. Thank you all!

**Comment by fnichol** [comment]
Very sorry about this.
I'll close this out but if I'm incorrect, please open/reopen.
